File: backend/src/schemas/emissions.py
"""Pydantic response schemas for the /emissions endpoints.

These mirror the contract defined in contracts/emissions-api.md.
"""

from __future__ import annotations

from pydantic import BaseModel, Field


# ── Shared primitives ─────────────────────────────────────────────────────────

class ScopeEntry(BaseModel):
    co2e_tonnes: float
    percentage: float


class ByScopeBreakdown(BaseModel):
    scope_1: ScopeEntry
    scope_2: ScopeEntry
    scope_3: ScopeEntry


class QuestionGroupEntry(BaseModel):
    question_group: str
    co2e_tonnes: float
    percentage: float


class TopSource(BaseModel):
    source_id: str
    activity: str
    question: str
    question_group: str
    scope: int
    co2e_tonnes: float
    percentage: float


# ── Overview ──────────────────────────────────────────────────────────────────

class EmissionsOverviewResponse(BaseModel):
    organisation_id: str
    organisation_name: str
    year: int
    market_factor_type: str
    total_co2e_tonnes: float
    by_scope: ByScopeBreakdown
    by_question_group: list[QuestionGroupEntry]
    top_sources: list[TopSource]
    available_years: list[int]


# ── Hierarchy ─────────────────────────────────────────────────────────────────

class HierarchyNodeResponse(BaseModel):
    id: str
    company_unit_id: int
    company_unit_name: str
    company_unit_type: str
    facility_type: str | None = None
    country: str | None = None
    country_code: str | None = None
    total_co2e_tonnes: float
    children: list[HierarchyNodeResponse] = Field(default_factory=list)

    model_config = {"from_attributes": True}


# Required for the self-referential model
HierarchyNodeResponse.model_rebuild()


class HierarchyResponse(BaseModel):
    root: HierarchyNodeResponse
    reporting_year: int


# ── Unit Detail ───────────────────────────────────────────────────────────────

class ChildUnitSummary(BaseModel):
    id: str
    company_unit_name: str
    company_unit_type: str
    total_co2e_tonnes: float


class RecordDetail(BaseModel):
    id: str
    year: int
    month: int
    scope: int
    market_factor_type: str
    value: float
    co2e_kg: float
    quality: str | None = None
    upstream: str
    upstream_name: str | None = None


class SourceDetail(BaseModel):
    id: str
    answer_id: int
    activity: str
    question: str
    question_group: str
    answer_unit: str
    co2e_tonnes: float
    scopes: list[int]
    records: list[RecordDetail]


class UnitDetailResponse(BaseModel):
    id: str
    company_unit_id: int
    company_unit_name: str
    company_unit_type: str
    facility_type: str | None = None
    city: str | None = None
    country: str | None = None
    country_code: str | None = None
    market_factor_type: str
    total_co2e_tonnes: float
    by_scope: ByScopeBreakdown
    by_question_group: list[QuestionGroupEntry]
    top_sources: list[TopSource]
    child_units: list[ChildUnitSummary]
    sources: list[SourceDetail]


# ── Trends ────────────────────────────────────────────────────────────────────

class TrendPoint(BaseModel):
    year: int
    co2e_tonnes: float


class TrendsResponse(BaseModel):
    unit_id: str | None
    company_unit_name: str
    market_factor_type: str
    trends: list[TrendPoint]


# ── Sources list (for initiative target selection) ────────────────────────────

class SourceSummary(BaseModel):
    id: str
    answer_id: int
    activity: str
    question: str
    question_group: str
    answer_unit: str
    company_unit_id: str
    company_unit_name: str
    company_unit_type: str
    co2e_tonnes: float
    scopes: list[int]


class SourceListResponse(BaseModel):
    sources: list[SourceSummary]

File: frontend/src/types/emissions.ts
// ─── Emissions API TypeScript types ───────────────────────────────────────────
// Mirrors contracts/emissions-api.md response schemas

export type MarketFactorType = 'Location' | 'Market';

// ── Overview ──────────────────────────────────────────────────────────────────

export interface ScopeEntry {
  co2e_tonnes: number;
  percentage: number;
}

export interface ByScopeBreakdown {
  scope_1: ScopeEntry;
  scope_2: ScopeEntry;
  scope_3: ScopeEntry;
}

export interface QuestionGroupEntry {
  question_group: string;
  co2e_tonnes: number;
  percentage: number;
}

export interface TopSource {
  source_id: string;
  activity: string;
  question: string;
  question_group: string;
  scope: 1 | 2 | 3;
  co2e_tonnes: number;
  percentage: number;
}

export interface EmissionsOverview {
  organisation_id: string;
  organisation_name: string;
  year: number;
  market_factor_type: MarketFactorType;
  total_co2e_tonnes: number;
  by_scope: ByScopeBreakdown;
  by_question_group: QuestionGroupEntry[];
  top_sources: TopSource[];
  available_years: number[];
}

// ── Hierarchy ─────────────────────────────────────────────────────────────────

export interface HierarchyNode {
  id: string;
  company_unit_id: number;
  company_unit_name: string;
  company_unit_type: 'division' | 'site';
  facility_type: string | null;
  country: string | null;
  country_code: string | null;
  total_co2e_tonnes: number;
  children: HierarchyNode[];
}

export interface HierarchyResponse {
  root: HierarchyNode;
  reporting_year: number;
}

// ── Unit Detail ───────────────────────────────────────────────────────────────

export interface ChildUnitSummary {
  id: string;
  company_unit_name: string;
  company_unit_type: 'division' | 'site';
  total_co2e_tonnes: number;
}

export interface RecordDetail {
  id: string;
  year: number;
  month: number;
  scope: 1 | 2 | 3;
  market_factor_type: MarketFactorType;
  value: number;
  co2e_kg: number;
  quality: string | null;
  upstream: string;
  upstream_name: string | null;
}

export interface SourceDetail {
  id: string;
  answer_id: number;
  activity: string;
  question: string;
  question_group: string;
  answer_unit: string;
  co2e_tonnes: number;
  scopes: number[];
  quality?: 'Actual' | 'Estimated' | 'Missing';
  records: RecordDetail[];
}

export interface UnitDetail {
  id: string;
  company_unit_id: number;
  company_unit_name: string;
  company_unit_type: 'division' | 'site';
  facility_type: string | null;
  city: string | null;
  country: string | null;
  country_code: string | null;
  market_factor_type: MarketFactorType;
  total_co2e_tonnes: number;
  by_scope: ByScopeBreakdown;
  by_question_group: QuestionGroupEntry[];
  top_sources: TopSource[];
  child_units: ChildUnitSummary[];
  sources: SourceDetail[];
}

// ── Trends ────────────────────────────────────────────────────────────────────

export interface TrendPoint {
  year: number;
  co2e_tonnes: number;
}

export interface TrendData {
  unit_id: string | null;
  company_unit_name: string;
  market_factor_type: MarketFactorType;
  trends: TrendPoint[];
}

// ── Sources list ──────────────────────────────────────────────────────────────

export interface EmissionSourceSummary {
  id: string;
  answer_id: number;
  activity: string;
  question: string;
  question_group: string;
  answer_unit: string;
  company_unit_id: string;
  company_unit_name: string;
  company_unit_type: 'division' | 'site';
  co2e_tonnes: number;
  scopes: number[];
}

export interface SourceListResponse {
  sources: EmissionSourceSummary[];
}
